### PR TITLE
[lldb] Add 'FindInMemory()' overload for PostMortemProcess.

### PR DESCRIPTION
**Pull request by mbucko**
Summary:
The current implementation of 'Process::FindInMemory()' utilizes a slow ReadMemory() API to search for 1 byte at a time in memory. This new overload takes advantage of the fact that the process core is already loaded into memory, allowing for a direct and more efficient search.

Test Plan:
llvm-lit ./llvm-project/lldb/test/API/python_api/find_in_memory/TestFindInMemoryCore.py ./llvm-project/lldb/test/API/python_api/find_in_memory/TestFindRangesInMemory.py ./llvm-project/lldb/test/API/python_api/find_in_memory/TestFindInMemory.py

### CODE REVIEW

File: lldb/source/Plugins/Process/elf-core/ProcessElfCore.cpp [hunk 1]
```diff
@@ -1095,3 +1095,9 @@ bool ProcessElfCore::GetProcessInfo(ProcessInstanceInfo &info) {
   }
   return true;
 }
+
+const uint8_t *ProcessElfCore::PeekMemory(lldb::addr_t low, lldb::addr_t high,
```
Comment (Jlalond) [inline lldb/source/Plugins/Process/elf-core/ProcessElfCore.cpp:1099]: Would an `ArrayRef<uint8_t>` be better for reading from a core file?

File: lldb/source/Plugins/Process/elf-core/ProcessElfCore.h [hunk 2]
```diff
@@ -111,6 +111,9 @@ class ProcessElfCore : public lldb_private::PostMortemProcess {
 
   bool SupportsMemoryTagging() override { return !m_core_tag_ranges.IsEmpty(); }
 
+  const uint8_t *PeekMemory(lldb::addr_t low, lldb::addr_t high,
```
Comment (Jlalond) [inline lldb/source/Plugins/Process/elf-core/ProcessElfCore.h:114]: Same ArrayRef<uint8_t> preference

File: lldb/include/lldb/Target/Process.h [hunk 3]
```diff
@@ -2835,6 +2835,34 @@ void PruneThreadPlans();
...
+        j--;
+      if (j < 0)
+        return low + s;
+      else
+        s += bad_char_heuristic[iterator[s + size - 1]];
+    }
+
+    return LLDB_INVALID_ADDRESS;
+  }
+
```
Comment (mbucko) [inline lldb/include/lldb/Target/Process.h:2865]: [NFC] Moved from 'Process::FindInMemory()'

File: lldb/include/lldb/Symbol/ObjectFile.h [hunk 4]
```diff
@@ -669,6 +669,13 @@ class ObjectFile : public std::enable_shared_from_this<ObjectFile>,
   // transparent decompression of section contents.
   size_t CopyData(lldb::offset_t offset, size_t length, void *dst) const;
 
+  // Returns a pointer to the data at the specified offset. If the offset is
+  // invalid, this function will return nullptr. The 'available_bytes' argument
+  // will be set to the number of bytes available at the given offset, which
+  // will be at most 'length' bytes.
+  const uint8_t *PeekData(lldb::addr_t offset, size_t length,
+                          size_t &available_bytes) const;
+
```
Comment (clayborg) [inline lldb/include/lldb/Symbol/ObjectFile.h:678]: return `llvm::ArrayRef<uint8_t>` and get rid of the `size_t &available_bytes` parameter as the size of the returned ArrayRef will be sufficient. Also, please provide a default implementation that returns an empty array so subclasses that don't override this don't have to have change to their header and implementation file.

File: lldb/include/lldb/Target/ProcessTrace.h [hunk 5]
```diff
@@ -72,6 +72,9 @@ class ProcessTrace : public PostMortemProcess {
   bool DoUpdateThreadList(ThreadList &old_thread_list,
                           ThreadList &new_thread_list) override;
 
+  const uint8_t *PeekMemory(lldb::addr_t low, lldb::addr_t high,
+                            size_t &available_bytes) override;
+
```
Comment (clayborg) [inline lldb/include/lldb/Target/ProcessTrace.h:77]: revert and use default implementation from Process.h as suggested.

File: lldb/include/lldb/Target/PostMortemProcess.h [hunk 6]
```diff
@@ -33,6 +34,23 @@ class PostMortemProcess : public Process {
...
 
 protected:
+  typedef lldb_private::Range<lldb::addr_t, lldb::addr_t> FileRange;
+  typedef lldb_private::RangeDataVector<lldb::addr_t, lldb::addr_t, FileRange>
+      VMRangeToFileOffset;
+  typedef lldb_private::RangeDataVector<lldb::addr_t, lldb::addr_t, uint32_t>
+      VMRangeToPermissions;
+
+  virtual const uint8_t *PeekMemory(lldb::addr_t low, lldb::addr_t high,
+                                    size_t &size) = 0;
```
Comment (clayborg) [inline lldb/include/lldb/Target/PostMortemProcess.h:44]: return `llvm::ArrayRef<uint8_t>` and get rid of the `size_t &available_bytes` parameter as the size of the returned ArrayRef will be sufficient. Do we want to return a `std::optional<llvm::ArrayRef<uint8_t>>` where if something comes back with no value, then we know PeekMemory is not implemented and we can fall back to the original code?

File: lldb/include/lldb/Target/Process.h (hunk 3 above)
Comment (clayborg) [inline lldb/include/lldb/Target/Process.h:2865]: Does this need to be templatized? Can't we leave this in Process::FindInMemory() as the default implementation?

File: lldb/source/Plugins/Process/elf-core/ProcessElfCore.cpp (hunk 1 above)
Comment (clayborg) [reply lldb/source/Plugins/Process/elf-core/ProcessElfCore.cpp:1099]: I commented in the header file, ArrayRef is a good idea. We might want a `std::optional<ArrayRef<uint8_t>>` where if something with no value is returned, it means that PeekMemory is not implemented and we should fall back to the original `Process::FindInMemory`

File: lldb/include/lldb/Target/Process.h (hunk 3 above)
Comment (mbucko) [reply lldb/include/lldb/Target/Process.h:2865]: Yes, there are two usecases.Process uses ProcessMemoryIterator class as the iterator which then uses m_process.ReadMemory() to access the data. Then, the PostMortemProcess uses a raw pointer to the memory

**Review by clayborg — COMMENTED** [review]
Thinking about this some more here are my thoughts: both `ObjectFile::PeekData` and `PostMoretemProcess::PeekMemory` should have a way to say "I am not implemented".  Using either a llvm::Expected or std::optional. If `ObjectFile::PeekData` returns unimplemented, we need to fall back to `ObjectFile::CopyData`. If `PostMoretemProcess::PeekMemory` returns unimplemented, we need to fall back to standard `PostMoretemProcess::ReadMemory`. The main reason being: some core files are too large to mmap into memory and at some point we need to be able to not load the entire core file at once. If we don't load the entire core file, then we need to allow `PostMoretemProcess::PeekMemory` to succeed for areas it does have mapped into the process, and fail for areas where it can't. If we build in the fallback stuff as mentioned, then this all continues to work.

So the main idea is: if either ObjectFile or PostMortemProcess can return a reference to the data in the Peek calls, then should as long as they have this data mapped into the LLDB address space and can hand out a pointer to the data. If they can't, we should fall back to something that can copy the data into a buffer as mentioned with `CopyData` or `ReadMemory`. Core files or object files might compress up their data, in which case we wouldn't be able to hande out a buffer from calls to Peek, but we would need to unzip the buffer into the buffer supplied by CopyData and ReadMemory on the fly.

File: lldb/include/lldb/Target/PostMortemProcess.h (hunk 6 above)
Comment (mbucko) [reply lldb/include/lldb/Target/PostMortemProcess.h:44]: It is sufficient to return an empty ArrayRef. Also, that is what all other APIs use in lldb. I will keep it the return type as llvm::ArrayRef<uint8_t> but let me know if you do want me to use std::optional.

**Comment by mbucko** [comment]
> Thinking about this some more here are my thoughts: both `ObjectFile::PeekData` and `PostMoretemProcess::PeekMemory` should have a way to say "I am not implemented". Using either a llvm::Expected or std::optional. If `ObjectFile::PeekData` returns unimplemented, we need to fall back to `ObjectFile::CopyData`. If `PostMoretemProcess::PeekMemory` returns unimplemented, we need to fall back to standard `PostMoretemProcess::ReadMemory`. The main reason being: some core files are too large to mmap into memory and at some point we need to be able to not load the entire core file at once. If we don't load the entire core file, then we need to allow `PostMoretemProcess::PeekMemory` to succeed for areas it does have mapped into the process, and fail for areas where it can't. If we build in the fallback stuff as mentioned, then this all continues to work.
> 
> So the main idea is: if either ObjectFile or PostMortemProcess can return a reference to the data in the Peek calls, then should as long as they have this data mapped into the LLDB address space and can hand out a pointer to the data. If they can't, we should fall back to something that can copy the data into a buffer as mentioned with `CopyData` or `ReadMemory`. Core files or object files might compress up their data, in which case we wouldn't be able to hande out a buffer from calls to Peek, but we would need to unzip the buffer into the buffer supplied by CopyData and ReadMemory on the fly.

This is already the behavior, if ObjectFile::PeekData returns anything other than an array with expected size, we fall back to the default implementation which uses the original ReadMemory() method.

File: lldb/include/lldb/Target/PostMortemProcess.h (hunk 6 above)
Comment (clayborg) [reply lldb/include/lldb/Target/PostMortemProcess.h]: > It is sufficient to return an empty ArrayRef. Also, that is what all other APIs use in lldb. I will keep it the return type as llvm::ArrayRef<uint8_t> but let me know if you do want me to use std::optional.

If returning a empty `llvm::ArrayRef<uint8_t>` is enough for us to be able to fall back to `ReadMemory` then this is fine.

**Review by clayborg — COMMENTED** [review]
Looks pretty close to me. Lets give some time for others to review

File: lldb/source/Symbol/ObjectFile.cpp [hunk 7]
```diff
@@ -482,6 +482,13 @@ size_t ObjectFile::CopyData(lldb::offset_t offset, size_t length,
   return m_data.CopyData(offset, length, dst);
 }
 
+llvm::ArrayRef<uint8_t> ObjectFile::PeekData(lldb::addr_t offset,
+                                             size_t length) const {
+  const uint8_t *data = m_data.PeekData(offset, length);
+  llvm::ArrayRef<uint8_t> array(data, length);
+  return array;
+}
+
```
Comment (labath) [inline lldb/source/Symbol/ObjectFile.cpp:491]: Could we avoid adding this function? You should be able access the same data using the `DataExtractor` overload of `GetData` (line 471, just call `GetData` on the returned DataExtractor), and this function is basically unimplementable for in-memory object files, which read their data on demand (as there's noone who can own the returned data).

(I realize this PR is not interested in those, but this API is going to be available for all object files)

**Comment by labath** [comment]
It just so happens that I got a `memory find` bug today, which made me realize just how broken the current behavior is -- it will effectively abort the search as soon as it runs into any unreadable block of memory (plus it also reads one byte at a time, which makes it excruciatingly slow for large ranges).

Now, this patch fixes both of those problems for core files (which implement the required APIs), but it still leaves the operation for live processes in its current, broken, state. This got me thinking whether it would be possible to implement something like this PR in a way that benefits all implementations -- and I think the answer to that question is "yes".

The part where you skip over the holes in memory is implementable using the existing GetMemoryRegion API which all processes implement (or should implement).

And we can certainly do something to avoid copying in the ReadMemory calls. Probably the best think to do would be to have a `ReadMemory` overload which returns a DataExtractor. This will give individual implementations the choice of how to return the memory. Implementations (like the core files, but maybe not all of them) which have the memory available already, can just return a DataExtractor referencing existing memory, while others (most live processes) can create a new DataBuffer object to back the DataExtractor.

I think that would give us most of the performance of the current implementation for core files, and also significantly improve the live process situation. (And we'd also have a single implementation).

What to you (all) think? Note that this isn't something I expect you must do on your own. Since this is something that I now need to do on my own anyway, I'd be happy to do some or all of this work myself.

**Comment by labath** [comment]
> So the main idea is: if either ObjectFile or PostMortemProcess can return a reference to the data in the Peek calls, then should as long as they have this data mapped into the LLDB address space and can hand out a pointer to the data. If they can't, we should fall back to something that can copy the data into a buffer as mentioned with `CopyData` or `ReadMemory`. Core files or object files might compress up their data, in which case we wouldn't be able to hande out a buffer from calls to Peek, but we would need to unzip the buffer into the buffer supplied by CopyData and ReadMemory on the fly.

BTW, we have another bug where the person wants to use lldb to open `/proc/kcore` (which is a special file, which pretends to be a (ELF) core file, but actually points to the live memory of the system). This core file cannot be mmapped, because a) the kernel won't let us, b) it's as big as the virtual address space (128TB for me) of the kernel. It works fine with gdb, but lldb chokes when it tries to mmap it.

Anyway, I'm not going to work on this any time soon, just thought you might want to hear about another use case where its not possible to return a reference to existing data.

**Comment by mbucko** [comment]
> It just so happens that I got a `memory find` bug today, which made me realize just how broken the current behavior is -- it will effectively abort the search as soon as it runs into any unreadable block of memory (plus it also reads one byte at a time, which makes it excruciatingly slow for large ranges).
> 
> Now, this patch fixes both of those problems for core files (which implement the required APIs), but it still leaves the operation for live processes in its current, broken, state. This got me thinking whether it would be possible to implement something like this PR in a way that benefits all implementations -- and I think the answer to that question is "yes".
> 
> The part where you skip over the holes in memory is implementable using the existing GetMemoryRegion API which all processes implement (or should implement).
> 
> And we can certainly do something to avoid copying in the ReadMemory calls. Probably the best think to do would be to have a `ReadMemory` overload which returns a DataExtractor. This will give individual implementations the choice of how to return the memory. Implementations (like the core files, but maybe not all of them) which have the memory available already, can just return a DataExtractor referencing existing memory, while others (most live processes) can create a new DataBuffer object to back the DataExtractor.
> 
> I think that would give us most of the performance of the current implementation for core files, and also significantly improve the live process situation. (And we'd also have a single implementation).
> 
> What to you (all) think? Note that this isn't something I expect you must do on your own. Since this is something that I now need to do on my own anyway, I'd be happy to do some or all of this work myself.

This would be quite the rework of the FindInMemory() API. Would it be worth creating an Issue and move the discussion there to unblock this PR?

**Comment by labath** [comment]
> This would be quite the rework of the FindInMemory() API. Would it be worth creating an Issue and move the discussion there to unblock this PR?

I think the two are related because if we (I?) do something like I proposed, then this patch becomes unnecessary (or even gets in the way).

I also don't think that would be as much work as you think. In fact, if we look at the number of new APIs added, I'm certain that implementation would end up being simpler.  `GetMemoryRegionInfo` is an existing API, and it serves a similar purpose as `VMRangeToFileOffset` in this patch, so the change would mainly consist of rewriting `FindInMemory` to use that instead. The only new API being added is the DataExtractor version of the `ReadMemory` function, but that one can be naturally split into a separate patch (and it may not even be needed -- depending on exactly how much performance you want to get out of this).

File: lldb/source/Target/PostMortemProcess.cpp [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+llvm::ArrayRef<uint8_t>
+PostMortemProcess::DoPeekMemory(lldb::ModuleSP &core_module_sp,
+                                VMRangeToFileOffset &core_aranges,
+                                lldb::addr_t low, lldb::addr_t high) {
+
+  ObjectFile *core_objfile = core_module_sp->GetObjectFile();
+
+  if (core_objfile == nullptr) {
+    return {};
+  }
```
Comment (clayborg) [inline lldb/source/Target/PostMortemProcess.cpp:53]: remove {} from single statement if per llvm coding guidelines.

File: lldb/source/Target/PostMortemProcess.cpp [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+
+  const VMRangeToFileOffset::Entry *core_memory_entry =
+      core_aranges.FindEntryThatContains(low);
+  if (core_memory_entry == nullptr || core_memory_entry->GetRangeEnd() < low) {
+    return {};
+  }
```
Comment (clayborg) [inline lldb/source/Target/PostMortemProcess.cpp:59]: remove {} from single statement if per llvm coding guidelines

File: lldb/source/Target/PostMortemProcess.cpp [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+  const lldb::addr_t offset = low - core_memory_entry->GetRangeBase();
+  const lldb::addr_t file_start = core_memory_entry->data.GetRangeBase();
+  const lldb::addr_t file_end = core_memory_entry->data.GetRangeEnd();
+
+  if (file_start == file_end) {
+    return {};
+  }
```
Comment (clayborg) [inline lldb/source/Target/PostMortemProcess.cpp:66]: remove {} from single statement if per llvm coding guidelines

File: lldb/source/Target/PostMortemProcess.cpp [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+  size_t bytes_available = 0;
+  if (file_end > file_start + offset)
+    bytes_available = file_end - (file_start + offset);
+
+  size_t bytes_to_read = high - low;
+  bytes_to_read = std::min(bytes_to_read, bytes_available);
+  if (bytes_to_read == 0) {
+    return {};
+  }
```
Comment (clayborg) [inline lldb/source/Target/PostMortemProcess.cpp:75]: remove {} from single statement if per llvm coding guidelines

File: lldb/source/Target/PostMortemProcess.cpp [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+  DataExtractor extractor;
+  core_objfile->GetData(core_memory_entry->data.GetRangeBase() + offset,
+                        bytes_to_read, extractor);
+  if (extractor.GetByteSize() != bytes_to_read) {
+    return {};
+  }
```
Comment (clayborg) [inline lldb/source/Target/PostMortemProcess.cpp:81]: remove {} from single statement if per llvm coding guidelines

**Review by clayborg — CHANGES_REQUESTED** [review]

**Review by clayborg — COMMENTED** [review]
I vote to leave this PR as is because `VMRangeToFileOffset` is a list of VMRanges that map to file offsets and this list combines entries with consecutive address ranges whose data is consecutive in the core file. And `VMRangeToPermissions` does not break things up. So not sure how we could use memory regions with this. It would also make the patch really large to have `Process::ReadMemory` return a `DataExtractor` and change all code over to using that. I do like this idea and think that we should work on a patch that can do this because `PeekMemory` could easily set the data pointers within a `DataExtractor` without copying the data.

**Comment by clayborg** [comment]
That being said, if there is something we can do to this PR to make it easy for you (Pavel) to implement this feature, we can make those changes. Like I think it would be fair to change all `PeekMemory/DoPeekMemory` over a virtual `ReadMemory/DoReadMemory` that returns a `DataExtractor` as this can easily take place of the `ArrayRef<uint8_t>` values that are returned.

This:
```
  virtual llvm::ArrayRef<uint8_t> PeekMemory(lldb::addr_t low, lldb::addr_t high);
```
Could become:
```
  virtual DataExtractor ReadMemory(lldb::addr_t low, lldb::addr_t high);
```
And we can adjust all users of `PeekMemory` and `DoPeakMemory` to use the the above functions?

**Comment by labath** [comment]
> That being said, if there is something we can do to this PR to make it easy for you (Pavel) to implement this feature, we can make those changes. Like I think it would be fair to change all `PeekMemory/DoPeekMemory` over a virtual `ReadMemory/DoReadMemory` that returns a `DataExtractor` as this can easily take place of the `ArrayRef<uint8_t>` values that are returned.
> 
> This:
> 
> ```
>   virtual llvm::ArrayRef<uint8_t> PeekMemory(lldb::addr_t low, lldb::addr_t high);
> ```
> 
> Could become:
> 
> ```
>   virtual DataExtractor ReadMemory(lldb::addr_t low, lldb::addr_t high);
> ```
> 
> And we can adjust all users of `PeekMemory` and `DoPeakMemory` to use the the above functions?

I'm not sure which feature are you referring to now. I'm not particularly worried about the `/proc/kcore` use case, as I don't know if I'll ever get to that, and it will require more substantial changes anyway.

The thing I am looking into right now is the ability to make `memory find` better in general, and not just for core files. My thinking is that if I can implement (and I'm offering to do that) a better generic implementation of `memory find` then probably the entirety of this patch becomes redundant. But to know that for sure, I'd like to understand more about your motivation for it. Like, what is the main problem that you're trying to solve with this? Is it to fix some specific bug (my bug was that the operation aborts as soon as it encounters a hole in the address space)? Or is it to make it faster? If so, what level of performance would you consider satisfactory?

**Comment by mbucko** [comment]
> > That being said, if there is something we can do to this PR to make it easy for you (Pavel) to implement this feature, we can make those changes. Like I think it would be fair to change all `PeekMemory/DoPeekMemory` over a virtual `ReadMemory/DoReadMemory` that returns a `DataExtractor` as this can easily take place of the `ArrayRef<uint8_t>` values that are returned.
> > This:
> > ```
> >   virtual llvm::ArrayRef<uint8_t> PeekMemory(lldb::addr_t low, lldb::addr_t high);
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Could become:
> > ```
> >   virtual DataExtractor ReadMemory(lldb::addr_t low, lldb::addr_t high);
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > And we can adjust all users of `PeekMemory` and `DoPeakMemory` to use the the above functions?
> 
> I'm not sure which feature are you referring to now. I'm not particularly worried about the `/proc/kcore` use case, as I don't know if I'll ever get to that, and it will require more substantial changes anyway.
> 
> The thing I am looking into right now is the ability to make `memory find` better in general, and not just for core files. My thinking is that if I can implement (and I'm offering to do that) a better generic implementation of `memory find` then probably the entirety of this patch becomes redundant. But to know that for sure, I'd like to understand more about your motivation for it. Like, what is the main problem that you're trying to solve with this? Is it to fix some specific bug (my bug was that the operation aborts as soon as it encounters a hole in the address space)? Or is it to make it faster? If so, what level of performance would you consider satisfactory?

The purpose of this it to make the `find memory` faster for post mortem processes. This patch gives us nearly 100x speed up based on my tests.

**Comment by clayborg** [comment]
Yes, this is primarily for performance of C++ coroutine detection where we must search memory for magic values.

**Comment by labath** [comment]
> The purpose of this it to make the `find memory` faster for post mortem processes. This patch gives us nearly 100x speed up based on my tests.

I see. And do you have an idea of how load-bearing that number is? For example, would you settle for a 18x speedup? That's what I got from #104193, with the advantage that it also works for live processes (with a ~6x speedup) and requires no new APIs.

With the addition of a copy-free ReadMemory API it could be made even faster (probably up to your 100x).

**Comment by jasonmolenda** [comment]
I don't want to derail the conversation about this PR, but I've been wanting to overhaul our Target::ReadMemory and Process::ReadMemory API in lldb for months now, and haven't put together a coherent proposal for it yet.  I need to clear a few things off my task list and get started on this for real.  My main complaint about the existing API is that many people don't understand the difference between a Target memory read (may read from live memory, may read from the file as an optimization, or if the memory is inaccessible) versus a Process memory read (only read from live memory), and the fact that Process has several convenient access methods like ReadCStringFromMemory which Target doesn't have means people chose based on that.

I haven't thought about all the different parameters we need to allow people to specify yet, and how it's going to look in an API, but I think we need an API that doesn't depend on the Target/Process distinction to implicitly get one behavior over another.  I hope to think on this enough to get a proposal on a Discourse soon.

(and so making changes to how the ReadMemory API return their data would make sense, since everything will need to be updated anyway.)

**Comment by labath** [comment]
Thanks for chiming in Jason. I'm glad this is getting attention, as I think that our memory reading APIs are in need of an overhaul -- and it gives me a chance to share my idea. ;)

I don't think that we need to change all of the APIs to return their memory. Sometimes reading the memory into a caller-provided buffer makes perfect sense. What I would like to do is to regularize all the APIs, and avoid the need to implement the logic of e.g. "reading a c string twice". The approach I'd like is to define a `MemoryReader` interface, which would contain all of the ways one may want to read a memory (as a string, integer, etc..). It would also contain a default implementation of the APIs so that an implementation would only need to implement one or two methods and have everything fall out naturally. Then, anything that wants to support that interface, can just inherit from it and implement the required method. If something wants to provide two different methods for reading the memory (I believe Process supports cached and uncached reads), it can just do two implementations, and provide them via different accessors (e.g. `process->CachedMemory()->ReadInt(addr)` and process->DirectMemory()->ReadInt(addr)`).

Another nice part about this is that if we have some code, which only needs to read memory, then it can take a MemoryReader reference (instead of Process or whatever), and then it can be unit tested by just passing it a mock memory reader.

**Comment by mbucko** [comment]
Let me know if you guys don't want this patch in. I will closed it and apply it to our local branch.

**Comment by jasonmolenda** [comment]
My comment about hoping to crack open the Target/Process ReadMemory API shouldn't have anything to do with this patch, I don't think.  I saw the discussion going in to areas of "the ReadMemory API could behave differently", and I wanted to note that I'm hoping to start a proposal on changing these API, and that would be a good opportunity to make changes like those because I'll probably need to update all the callers across the codebase if we find a new API design we can all agree on.

**Comment by mbucko** [comment]
> Let me know if you guys don't want this patch in. I will closed it and apply it to our local branch. @jasonmolenda @labath

@labath just a friendly ping

**Comment by labath** [comment]
> Let me know if you guys don't want this patch in. I will closed it and apply it to our local branch. @jasonmolenda @labath

Without claiming official authority to make this decision, I'm going to say that *I* don't think this is a good idea, because the patch is very (and IMO, unnecessarily) specific to post mortem processes. My (currently WIP) patch (#104193) gets you most of the speed benefits, works with live processes as well and it does that without introducing any new APIs.

What you do downstream is up to you, but I would encourage you to wait for me to finish up that patch, as this doesn't look like something that would be a good idea to carry downstream (I certainly wouldn't want to do it).

**Comment by mbucko** [comment]
Closing this PR in favor of #104193

**Pull request closed without merging**